### ISSUE TITLE
matchExactly matcher for Maps?

### ISSUE THREAD

**Issue by tlinkowski** [kotest/kotest#3246]
## Problem

I would like to test if a `Map`:
- has key-value pairs where values match custom assertions (like `matchAll`)
- doesn't have any extra keys (like `containsExactly`)

## Solution

I think the `Matcher` implementation handling `matchAll` could be adapted like `MapContainsMatcher` was (`private val ignoreExtraKeys: Boolean = false`). This way, the implementation could serve both existing `matchAll` and new `matchExactly`.

## Workaround

I worked around this problem with the following "dirty" code:
```kotlin
inline fun <reified K, V> Map<K, V>.shouldMatchExactly(vararg matchers: Pair<K, (V) -> Unit>) {
    shouldMatchAll(*matchers)
    val extraKeys = keys - matchers.map { it.first }.toSet()
    if (extraKeys.isNotEmpty()) {
        shouldNotHaveKeys(*extraKeys.toTypedArray())
    }
}
```

## Documentation

PS. The "no extra" bit in the documentation of `shouldMatchAll` seems incorrect:

Maps | &nbsp;
-- | --
`map.shouldMatchAll("k1" to {it shouldBe "v1"}, "k2" to {it shouldBe "v2"}, ...)` | Asserts that all the entries in the map can be matched with the provided matchers, and _no extra_.

But that's exactly what I'd like to achieve with `shouldMatchExactly`.

https://kotest.io/docs/assertions/core-matchers.html

**Comment by Kantis** [comment]
I think this is a good idea

**Comment by pientaa** [comment]
@Kantis shouldn't it be closed? It seems to be done here #3305

**Comment by Kantis** [comment]
@pientaa you're right. Don't remember why I reopened
